Source/Core/DolphinQt/GCMemcardManager.cpp

```diff
@@ -293,162 +344,314 @@ void GCMemcardManager::SetSlotFileInteractive(int slot)
       this,
       slot == 0 ? tr("Set memory card file for Slot A") : tr("Set memory card file for Slot B"),
       QString::fromStdString(File::GetUserPath(D_GCUSER_IDX)),
-      tr("GameCube Memory Cards (*.raw *.gcp)") + QStringLiteral(";;") + tr("All Files (*)")));
+      QStringLiteral("%1 (*.raw *.gcp);;%2 (*)")
+          .arg(tr("GameCube Memory Cards"), tr("All Files"))));
   if (!path.isEmpty())
     m_slot_file_edit[slot]->setText(path);
 }
 
-void GCMemcardManager::ExportFiles(bool prompt)
+std::vector<u8> GCMemcardManager::GetSelectedFileIndices()
 {
-  auto selection = m_slot_table[m_active_slot]->selectedItems();
-  auto& memcard = m_slot_memcard[m_active_slot];
-
-  auto count = selection.count() / m_slot_table[m_active_slot]->columnCount();
-
-  for (int i = 0; i < count; i++)
+  const auto selection = m_slot_table[m_active_slot]->selectedItems();
+  std::vector<bool> lookup(Memcard::DIRLEN);
+  for (const auto* item : selection)
   {
-    auto sel = selection[i * m_slot_table[m_active_slot]->columnCount()];
-    int file_index = memcard->GetFileIndex(m_slot_table[m_active_slot]->row(sel));
-
-    std::string gci_filename;
-    if (!memcard->GCI_FileName(file_index, gci_filename))
-      return;
-
-    QString path;
-    if (prompt)
+    const int index = item->data(Qt::UserRole).toInt();
+    if (index < 0 || index >= static_cast<int>(Memcard::DIRLEN))
     {
-      path = QFileDialog::getSaveFileName(
-          this, tr("Export Save File"),
-          QString::fromStdString(File::GetUserPath(D_GCUSER_IDX)) +
-              QStringLiteral("/%1").arg(QString::fromStdString(gci_filename)),
-          tr("Native GCI File (*.gci)") + QStringLiteral(";;") +
-              tr("MadCatz Gameshark files(*.gcs)") + QStringLiteral(";;") +
-              tr("Datel MaxDrive/Pro files(*.sav)"));
-
-      if (path.isEmpty())
-        return;
-    }
-    else
-    {
-      path = QString::fromStdString(File::GetUserPath(D_GCUSER_IDX)) +
-             QStringLiteral("/%1").arg(QString::fromStdString(gci_filename));
+      ModalMessageBox::warning(this, tr("Error"),
+                               tr("Data inconsistency in GCMemcardManager, aborting action."));
+      return {};
     }
+    lookup[index] = true;
+  }
 
-    // TODO: This is obviously intended to check for success instead.
-    const auto exportRetval = memcard->ExportGci(file_index, path.toStdString(), "");
-    if (exportRetval == Memcard::GCMemcardExportFileRetVal::UNUSED)
-    {
-      File::Delete(path.toStdString());
-    }
+  std::vector<u8> selected_indices;
+  for (u8 i = 0; i < Memcard::DIRLEN; ++i)
+  {
+    if (lookup[i])
+      selected_indices.push_back(i);
```

Comment:
Is there a reason you need two loops?

Reply:
Wish you had asked me this a few months ago, but the code as-written ensures that the returned indices never contain any duplicates, which is important. We also get a consistent ascending order of the indices, though I honestly don't remember if that part is important (probably not?). I suppose if we don't mind the arbitrary order we could do easily do it with one loop too -- maybe even preferable if that respects the order the user selected the items and/or the order of the items in the list?

Reply:
Ah yes, I didn't think about the ordering of the final result, that may matter.  No concerns with it as is, just was curious, thank you!